### PR TITLE
Fix InferType logic - add backward inference for some ops

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##
- Fixes partially https://github.com/apache/incubator-mxnet/issues/16757
- Also fixes: https://github.com/apache/incubator-mxnet/issues/16816

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/contrib/deformable_convolution-inl.h [hunk 1]
```diff
@@ -453,18 +454,10 @@ class DeformableConvolutionProp : public OperatorProperty {
...
-    CHECK_NE(dtype, -1) << "First input must have specified type";
-    for (size_t i = 0; i < in_type->size(); ++i) {
-      if ((*in_type)[i] == -1) {
-        (*in_type)[i] = dtype;
-      } else {
-        UNIFORM_TYPE_CHECK((*in_type)[i], dtype, ListArguments()[i]);
-      }
-    }
-    out_type->clear();
-    out_type->push_back(dtype);
```
Comment (ptrendx) [inline src/operator/contrib/deformable_convolution-inl.h]: Wow, this would just discard the type inferred by the other operators...

Comment (anirudh2290) [reply src/operator/contrib/deformable_convolution-inl.h]: yep, many operators behave this way. probably more than what we have captured in the issue list :)

Comment (ptrendx) [reply src/operator/contrib/deformable_convolution-inl.h]: Seems that the logic for type/shape inference in operators really needs to be looked at. I have another issue (and PR with a fix) coming today about how it is currently possible to not let the operator check their input and output attributes (and there should be also another how operators lie in their infertype/shape about whether they successfully inferred everything or not).

**Comment by anirudh2290** [comment]
closing this, since I am not working on this currently. If someone is interested in picking this up, please go ahead.

**Pull request closed without merging**